--- a/net/minecraft/world/entity/ai/behavior/BackUpIfTooClose.java
+++ b/net/minecraft/world/entity/ai/behavior/BackUpIfTooClose.java
@@ -8,7 +_,7 @@
 import net.minecraft.world.entity.ai.memory.NearestVisibleLivingEntities;
 
 public class BackUpIfTooClose {
-    public static OneShot<Mob> create(int tooCloseDistance, float strafeSpeed) {
+    public static synchronized OneShot<Mob> create(int tooCloseDistance, float strafeSpeed) {
         return BehaviorBuilder.create(
             i -> i.group(
                     i.absent(MemoryModuleType.WALK_TARGET),
@@ -17,7 +_,12 @@
                     i.present(MemoryModuleType.NEAREST_VISIBLE_LIVING_ENTITIES)
                 )
                 .apply(i, (walkTarget, lookTarget, attackTarget, nearestVisible) -> (level, body, timestamp) -> {
-                    LivingEntity target = i.get(attackTarget);
+                    LivingEntity target = BehaviorUtils.regionSafeEntity(body, i.get(attackTarget));
+                    if (target == null) {
+                        attackTarget.erase();
+                        return false;
+                    }
+
                     if (target.closerThan(body, tooCloseDistance) && i.<NearestVisibleLivingEntities>get(nearestVisible).contains(target)) {
                         lookTarget.set(new EntityTracker(target, true));
                         body.getMoveControl().strafe(-strafeSpeed, 0.0F);
