--- a/net/minecraft/world/entity/ai/behavior/MeleeAttack.java
+++ b/net/minecraft/world/entity/ai/behavior/MeleeAttack.java
@@ -9,11 +_,11 @@
 import net.minecraft.world.entity.ai.memory.NearestVisibleLivingEntities;
 
 public class MeleeAttack {
-    public static <T extends Mob> OneShot<T> create(int cooldownBetweenAttacks) {
+    public static synchronized <T extends Mob> OneShot<T> create(int cooldownBetweenAttacks) {
         return create(body -> true, cooldownBetweenAttacks);
     }
 
-    public static <T extends Mob> OneShot<T> create(Predicate<T> canAttackPredicate, int cooldownBetweenAttacks) {
+    public static synchronized <T extends Mob> OneShot<T> create(Predicate<T> canAttackPredicate, int cooldownBetweenAttacks) {
         return BehaviorBuilder.create(
             i -> i.group(
                     i.registered(MemoryModuleType.LOOK_TARGET),
@@ -24,7 +_,12 @@
                 .apply(
                     i,
                     (lookTarget, attackTarget, attackCoolingDown, nearestEntities) -> (level, body, timestamp) -> {
-                        LivingEntity target = i.get(attackTarget);
+                        LivingEntity target = BehaviorUtils.regionSafeEntity(body, i.get(attackTarget));
+                        if (target == null) {
+                            attackTarget.erase();
+                            return false;
+                        }
+
                         if (canAttackPredicate.test(body)
                             && !isHoldingUsableNonMeleeWeapon(body)
                             && body.isWithinMeleeAttackRange(target)
@@ -42,7 +_,7 @@
         );
     }
 
-    private static boolean isHoldingUsableNonMeleeWeapon(Mob body) {
+    private static synchronized boolean isHoldingUsableNonMeleeWeapon(Mob body) {
         return body.isHolding(body::canUseNonMeleeWeapon);
     }
 }
