--- a/net/minecraft/world/entity/ai/behavior/RandomLookAround.java
+++ b/net/minecraft/world/entity/ai/behavior/RandomLookAround.java
@@ -28,7 +_,7 @@
         }
     }
 
-    protected void start(ServerLevel level, Mob body, long timestamp) {
+    protected synchronized void start(ServerLevel level, Mob body, long timestamp) {
         RandomSource random = body.getRandom();
         float pitch = Mth.clamp(random.nextFloat() * this.pitchRange + this.minPitch, -90.0F, 90.0F);
         float rotation = Mth.wrapDegrees(body.getYRot() + 2.0F * random.nextFloat() * this.maxYaw - this.maxYaw);
