--- a/net/minecraft/world/entity/ai/behavior/StrollToPoiList.java
+++ b/net/minecraft/world/entity/ai/behavior/StrollToPoiList.java
@@ -9,7 +_,7 @@
 import org.apache.commons.lang3.mutable.MutableLong;
 
 public class StrollToPoiList {
-    public static BehaviorControl<Villager> create(
+    public static synchronized BehaviorControl<Villager> create(
         MemoryModuleType<List<GlobalPos>> strollToMemoryType,
         float speedModifier,
         int closeEnoughDist,
@@ -22,23 +_,25 @@
                 .apply(
                     i,
                     (walkTarget, strollToMemory, mustBeCloseToMemory) -> (level, body, timestamp) -> {
-                        List<GlobalPos> strollTo = i.get(strollToMemory);
-                        GlobalPos stayCloseTo = i.get(mustBeCloseToMemory);
-                        if (strollTo.isEmpty()) {
-                            return false;
-                        } else {
-                            GlobalPos targetPos = strollTo.get(level.getRandom().nextInt(strollTo.size()));
-                            if (targetPos != null
-                                && level.dimension() == targetPos.dimension()
-                                && stayCloseTo.pos().closerToCenterThan(body.position(), maxDistanceFromPoi)) {
-                                if (timestamp > nextOkStartTime.longValue()) {
-                                    walkTarget.set(new WalkTarget(targetPos.pos(), speedModifier, closeEnoughDist));
-                                    nextOkStartTime.setValue(timestamp + 100L);
-                                }
-
-                                return true;
-                            } else {
+                        synchronized (nextOkStartTime) {
+                            List<GlobalPos> strollTo = List.copyOf(i.get(strollToMemory));
+                            GlobalPos stayCloseTo = i.get(mustBeCloseToMemory);
+                            if (strollTo.isEmpty()) {
                                 return false;
+                            } else {
+                                GlobalPos targetPos = strollTo.get(body.getRandom().nextInt(strollTo.size()));
+                                if (targetPos != null
+                                    && level.dimension() == targetPos.dimension()
+                                    && stayCloseTo.pos().closerToCenterThan(body.position(), maxDistanceFromPoi)) {
+                                    if (timestamp > nextOkStartTime.longValue()) {
+                                        walkTarget.set(new WalkTarget(targetPos.pos(), speedModifier, closeEnoughDist));
+                                        nextOkStartTime.setValue(timestamp + 100L);
+                                    }
+
+                                    return true;
+                                } else {
+                                    return false;
+                                }
                             }
                         }
                     }
