--- a/net/minecraft/world/entity/ai/behavior/TriggerGate.java
+++ b/net/minecraft/world/entity/ai/behavior/TriggerGate.java
@@ -7,23 +_,25 @@
 import net.minecraft.world.entity.ai.behavior.declarative.Trigger;
 
 public class TriggerGate {
-    public static <E extends LivingEntity> OneShot<E> triggerOneShuffled(List<Pair<? extends Trigger<? super E>, Integer>> weightedTriggers) {
+    public static synchronized <E extends LivingEntity> OneShot<E> triggerOneShuffled(List<Pair<? extends Trigger<? super E>, Integer>> weightedTriggers) {
         return triggerGate(weightedTriggers, GateBehavior.OrderPolicy.SHUFFLED, GateBehavior.RunningPolicy.RUN_ONE);
     }
 
-    public static <E extends LivingEntity> OneShot<E> triggerGate(
+    public static synchronized <E extends LivingEntity> OneShot<E> triggerGate(
         List<Pair<? extends Trigger<? super E>, Integer>> weightedBehaviors, GateBehavior.OrderPolicy orderPolicy, GateBehavior.RunningPolicy runningPolicy
     ) {
         ShufflingList<Trigger<? super E>> behaviors = new ShufflingList<>();
-        weightedBehaviors.forEach(entry -> behaviors.add(entry.getFirst(), entry.getSecond()));
+        List.copyOf(weightedBehaviors).forEach(entry -> behaviors.add(entry.getFirst(), entry.getSecond()));
         return BehaviorBuilder.create(i -> i.point((level, body, timestamp) -> {
-            if (orderPolicy == GateBehavior.OrderPolicy.SHUFFLED) {
-                behaviors.shuffle();
-            }
+            synchronized (behaviors) {
+                if (orderPolicy == GateBehavior.OrderPolicy.SHUFFLED) {
+                    behaviors.shuffle();
+                }
 
-            for (Trigger<? super E> behavior : behaviors) {
-                if (behavior.trigger(level, body, timestamp) && runningPolicy == GateBehavior.RunningPolicy.RUN_ONE) {
-                    break;
+                for (Trigger<? super E> behavior : behaviors) {
+                    if (behavior.trigger(level, body, timestamp) && runningPolicy == GateBehavior.RunningPolicy.RUN_ONE) {
+                        break;
+                    }
                 }
             }
 
