--- a/net/minecraft/world/entity/ai/behavior/UpdateActivityFromSchedule.java
+++ b/net/minecraft/world/entity/ai/behavior/UpdateActivityFromSchedule.java
@@ -4,7 +_,7 @@
 import net.minecraft.world.entity.ai.behavior.declarative.BehaviorBuilder;
 
 public class UpdateActivityFromSchedule {
-    public static BehaviorControl<LivingEntity> create() {
+    public static synchronized BehaviorControl<LivingEntity> create() {
         return BehaviorBuilder.create(i -> i.point((level, body, timestamp) -> {
             body.getBrain().updateActivityFromSchedule(level.environmentAttributes(), level.getGameTime(), body.position());
             return true;
