--- a/net/minecraft/stats/ServerStatsCounter.java
+++ b/net/minecraft/stats/ServerStatsCounter.java
@@ -36,6 +_,7 @@
 import net.minecraft.util.Util;
 import net.minecraft.util.datafix.DataFixTypes;
 import net.minecraft.world.entity.player.Player;
+import net.neoforged.neoforge.server.region.RegionizedWorldGuard;
 import org.slf4j.Logger;
 
 public class ServerStatsCounter extends StatsCounter {
@@ -77,7 +_,7 @@
         }
     }
 
-    public void save() {
+    public synchronized void save() {
         try {
             FileUtil.createDirectoriesSafe(this.file.getParent());
 
@@ -90,18 +_,18 @@
     }
 
     @Override
-    public void setValue(Player player, Stat<?> stat, int count) {
+    public synchronized void setValue(Player player, Stat<?> stat, int count) {
         super.setValue(player, stat, count);
         this.dirty.add(stat);
     }
 
-    private Set<Stat<?>> getDirty() {
+    private synchronized Set<Stat<?>> getDirty() {
         Set<Stat<?>> result = Sets.newHashSet(this.dirty);
         this.dirty.clear();
         return result;
     }
 
-    public void parse(DataFixer fixerUpper, JsonElement element) {
+    public synchronized void parse(DataFixer fixerUpper, JsonElement element) {
         Dynamic<JsonElement> data = new Dynamic<>(JsonOps.INSTANCE, element);
         data = DataFixTypes.STATS.updateToCurrentVersion(fixerUpper, data, NbtUtils.getDataVersion(data, 1343));
         this.stats
@@ -112,14 +_,14 @@
             );
     }
 
-    protected JsonElement toJson() {
+    protected synchronized JsonElement toJson() {
         JsonObject result = new JsonObject();
-        result.add("stats", STATS_CODEC.encodeStart(JsonOps.INSTANCE, this.stats).getOrThrow());
+        result.add("stats", STATS_CODEC.encodeStart(JsonOps.INSTANCE, new Object2IntOpenHashMap<>(this.stats)).getOrThrow());
         result.addProperty("DataVersion", SharedConstants.getCurrentVersion().dataVersion().version());
         return result;
     }
 
-    public void markAllDirty() {
+    public synchronized void markAllDirty() {
         this.dirty.addAll(this.stats.keySet());
     }
 
@@ -130,6 +_,6 @@
             statsToSend.put(stat, this.getValue(stat));
         }
 
-        player.connection.send(new ClientboundAwardStatsPacket(statsToSend));
+        RegionizedWorldGuard.sendToPlayer(player, new ClientboundAwardStatsPacket(statsToSend));
     }
 }
