--- a/net/minecraft/world/entity/ai/goal/DoorInteractGoal.java
+++ b/net/minecraft/world/entity/ai/goal/DoorInteractGoal.java
@@ -7,6 +_,7 @@
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.pathfinder.Node;
 import net.minecraft.world.level.pathfinder.Path;
+import net.neoforged.neoforge.server.region.RegionizedWorldGuard;
 
 public abstract class DoorInteractGoal extends Goal {
     protected Mob mob;
@@ -23,9 +_,12 @@
         }
     }
 
-    protected boolean isOpen() {
+    protected synchronized boolean isOpen() {
         if (!this.hasDoor) {
             return false;
+        } else if (!this.canReadBlock(this.doorPos)) {
+            this.hasDoor = false;
+            return false;
         } else {
             BlockState blockState = this.mob.level().getBlockState(this.doorPos);
             if (!(blockState.getBlock() instanceof DoorBlock)) {
@@ -37,8 +_,8 @@
         }
     }
 
-    protected void setOpen(boolean open) {
-        if (this.hasDoor) {
+    protected synchronized void setOpen(boolean open) {
+        if (this.hasDoor && this.canReadBlock(this.doorPos)) {
             BlockState blockState = this.mob.level().getBlockState(this.doorPos);
             if (blockState.getBlock() instanceof DoorBlock) {
                 ((DoorBlock)blockState.getBlock()).setOpen(this.mob, this.mob.level(), blockState, this.doorPos, open);
@@ -47,7 +_,7 @@
     }
 
     @Override
-    public boolean canUse() {
+    public synchronized boolean canUse() {
         if (!GoalUtils.hasGroundPathNavigation(this.mob)) {
             return false;
         } else if (!this.mob.horizontalCollision) {
@@ -59,7 +_,7 @@
                     Node node = path.getNode(i);
                     this.doorPos = new BlockPos(node.x, node.y + 1, node.z);
                     if (!(this.mob.distanceToSqr(this.doorPos.getX(), this.mob.getY(), this.doorPos.getZ()) > 2.25)) {
-                        this.hasDoor = DoorBlock.isWoodenDoor(this.mob.level(), this.doorPos);
+                        this.hasDoor = this.canReadBlock(this.doorPos) && DoorBlock.isWoodenDoor(this.mob.level(), this.doorPos);
                         if (this.hasDoor) {
                             return true;
                         }
@@ -67,7 +_,7 @@
                 }
 
                 this.doorPos = this.mob.blockPosition().above();
-                this.hasDoor = DoorBlock.isWoodenDoor(this.mob.level(), this.doorPos);
+                this.hasDoor = this.canReadBlock(this.doorPos) && DoorBlock.isWoodenDoor(this.mob.level(), this.doorPos);
                 return this.hasDoor;
             } else {
                 return false;
@@ -76,29 +_,33 @@
     }
 
     @Override
-    public boolean canContinueToUse() {
+    public synchronized boolean canContinueToUse() {
         return !this.passed;
     }
 
     @Override
-    public void start() {
+    public synchronized void start() {
         this.passed = false;
         this.doorOpenDirX = (float)(this.doorPos.getX() + 0.5 - this.mob.getX());
         this.doorOpenDirZ = (float)(this.doorPos.getZ() + 0.5 - this.mob.getZ());
     }
 
     @Override
-    public boolean requiresUpdateEveryTick() {
+    public synchronized boolean requiresUpdateEveryTick() {
         return true;
     }
 
     @Override
-    public void tick() {
+    public synchronized void tick() {
         float newDoorDirX = (float)(this.doorPos.getX() + 0.5 - this.mob.getX());
         float newDoorDirZ = (float)(this.doorPos.getZ() + 0.5 - this.mob.getZ());
         float dot = this.doorOpenDirX * newDoorDirX + this.doorOpenDirZ * newDoorDirZ;
         if (dot < 0.0F) {
             this.passed = true;
         }
+    }
+
+    protected synchronized boolean canReadBlock(BlockPos pos) {
+        return RegionizedWorldGuard.shouldLoadChunk(this.mob.level(), pos.getX() >> 4, pos.getZ() >> 4);
     }
 }
