--- a/net/minecraft/world/entity/ai/goal/UseItemGoal.java
+++ b/net/minecraft/world/entity/ai/goal/UseItemGoal.java
@@ -22,23 +_,23 @@
     }
 
     @Override
-    public boolean canUse() {
+    public synchronized boolean canUse() {
         return this.canUseSelector.test(this.mob);
     }
 
     @Override
-    public boolean canContinueToUse() {
+    public synchronized boolean canContinueToUse() {
         return this.mob.isUsingItem();
     }
 
     @Override
-    public void start() {
+    public synchronized void start() {
         this.mob.setItemSlot(EquipmentSlot.MAINHAND, this.item.copy());
         this.mob.startUsingItem(InteractionHand.MAIN_HAND);
     }
 
     @Override
-    public void stop() {
+    public synchronized void stop() {
         this.mob.setItemSlot(EquipmentSlot.MAINHAND, ItemStack.EMPTY);
         if (this.finishUsingSound != null) {
             this.mob.playSound(this.finishUsingSound, 1.0F, this.mob.getRandom().nextFloat() * 0.2F + 0.9F);
