--- a/net/minecraft/world/entity/ai/goal/ZombieAttackGoal.java
+++ b/net/minecraft/world/entity/ai/goal/ZombieAttackGoal.java
@@ -12,19 +_,19 @@
     }
 
     @Override
-    public void start() {
+    public synchronized void start() {
         super.start();
         this.raiseArmTicks = 0;
     }
 
     @Override
-    public void stop() {
+    public synchronized void stop() {
         super.stop();
         this.zombie.setAggressive(false);
     }
 
     @Override
-    public void tick() {
+    public synchronized void tick() {
         super.tick();
         this.raiseArmTicks++;
         if (this.raiseArmTicks >= 5 && this.getTicksUntilNextAttack() < this.getAttackInterval() / 2) {
