--- a/net/minecraft/world/level/LevelReader.java
+++ b/net/minecraft/world/level/LevelReader.java
@@ -21,6 +_,7 @@
 import net.minecraft.world.level.dimension.DimensionType;
 import net.minecraft.world.level.levelgen.Heightmap;
 import net.minecraft.world.phys.AABB;
+import net.neoforged.neoforge.server.region.RegionizedWorldGuard;
 import org.jspecify.annotations.Nullable;
 
 public interface LevelReader extends BlockAndLightGetter, CollisionGetter, SignalGetter, BiomeManager.NoiseBiomeSource, net.neoforged.neoforge.common.extensions.ILevelReaderExtension {
@@ -130,6 +_,10 @@
 
     @Override
     default @Nullable BlockGetter getChunkForCollisions(int chunkX, int chunkZ) {
+        if (this instanceof Level level && !RegionizedWorldGuard.shouldLoadChunk(level, chunkX, chunkZ)) {
+            return null;
+        }
+
         return this.getChunk(chunkX, chunkZ, ChunkStatus.EMPTY, false);
     }
 
@@ -203,6 +_,10 @@
 
         for (int chunkX = chunkX0; chunkX <= chunkX1; chunkX++) {
             for (int chunkZ = chunkZ0; chunkZ <= chunkZ1; chunkZ++) {
+                if (this instanceof Level level && !RegionizedWorldGuard.shouldLoadChunk(level, chunkX, chunkZ)) {
+                    return false;
+                }
+
                 if (!this.hasChunk(chunkX, chunkZ)) {
                     return false;
                 }
