--- a/net/minecraft/world/entity/ai/control/BodyRotationControl.java
+++ b/net/minecraft/world/entity/ai/control/BodyRotationControl.java
@@ -15,7 +_,7 @@
         this.mob = mob;
     }
 
-    public void clientTick() {
+    public synchronized void clientTick() {
         if (this.isMoving()) {
             this.mob.yBodyRot = this.mob.getYRot();
             this.rotateHeadIfNecessary();
@@ -37,26 +_,26 @@
         }
     }
 
-    private void rotateBodyIfNecessary() {
+    private synchronized void rotateBodyIfNecessary() {
         this.mob.yBodyRot = Mth.rotateIfNecessary(this.mob.yBodyRot, this.mob.yHeadRot, this.mob.getMaxHeadYRot());
     }
 
-    private void rotateHeadIfNecessary() {
+    private synchronized void rotateHeadIfNecessary() {
         this.mob.yHeadRot = Mth.rotateIfNecessary(this.mob.yHeadRot, this.mob.yBodyRot, this.mob.getMaxHeadYRot());
     }
 
-    private void rotateHeadTowardsFront() {
+    private synchronized void rotateHeadTowardsFront() {
         int timeSinceStartingToFaceForward = this.headStableTime - 10;
         float faceForwardFraction = Mth.clamp(timeSinceStartingToFaceForward / 10.0F, 0.0F, 1.0F);
         float angleRemainingUntilFacingForward = this.mob.getMaxHeadYRot() * (1.0F - faceForwardFraction);
         this.mob.yBodyRot = Mth.rotateIfNecessary(this.mob.yBodyRot, this.mob.yHeadRot, angleRemainingUntilFacingForward);
     }
 
-    private boolean notCarryingMobPassengers() {
+    private synchronized boolean notCarryingMobPassengers() {
         return !(this.mob.getFirstPassenger() instanceof Mob);
     }
 
-    private boolean isMoving() {
+    private synchronized boolean isMoving() {
         double xd = this.mob.getX() - this.mob.xo;
         double zd = this.mob.getZ() - this.mob.zo;
         return xd * xd + zd * zd > 2.5000003E-7F;
