--- a/net/minecraft/world/entity/ai/control/SmoothSwimmingMoveControl.java
+++ b/net/minecraft/world/entity/ai/control/SmoothSwimmingMoveControl.java
@@ -23,7 +_,7 @@
     }
 
     @Override
-    public void tick() {
+    public synchronized void tick() {
         if (this.applyGravity && this.mob.isInWater()) {
             this.mob.setDeltaMovement(this.mob.getDeltaMovement().add(0.0, 0.005, 0.0));
         }
