--- a/net/minecraft/world/entity/ai/sensing/AxolotlAttackablesSensor.java
+++ b/net/minecraft/world/entity/ai/sensing/AxolotlAttackablesSensor.java
@@ -11,32 +_,32 @@
     public static final float TARGET_DETECTION_DISTANCE = 8.0F;
 
     @Override
-    protected boolean isMatchingEntity(ServerLevel level, LivingEntity body, LivingEntity mob) {
+    protected synchronized boolean isMatchingEntity(ServerLevel level, LivingEntity body, LivingEntity mob) {
         return this.isClose(body, mob)
             && mob.isInWater()
             && (this.isHostileTarget(mob) || this.isHuntTarget(body, mob))
             && Sensor.isEntityAttackable(level, body, mob);
     }
 
-    private boolean isHuntTarget(LivingEntity body, LivingEntity mob) {
+    private synchronized boolean isHuntTarget(LivingEntity body, LivingEntity mob) {
         return !body.getBrain().hasMemoryValue(MemoryModuleType.HAS_HUNTING_COOLDOWN) && mob.is(EntityTypeTags.AXOLOTL_HUNT_TARGETS);
     }
 
-    private boolean isHostileTarget(LivingEntity mob) {
+    private synchronized boolean isHostileTarget(LivingEntity mob) {
         return mob.is(EntityTypeTags.AXOLOTL_ALWAYS_HOSTILES);
     }
 
-    private boolean isClose(LivingEntity body, LivingEntity mob) {
+    private synchronized boolean isClose(LivingEntity body, LivingEntity mob) {
         return mob.distanceToSqr(body) <= 64.0;
     }
 
     @Override
-    protected MemoryModuleType<LivingEntity> getMemoryToSet() {
+    protected synchronized MemoryModuleType<LivingEntity> getMemoryToSet() {
         return MemoryModuleType.NEAREST_ATTACKABLE;
     }
 
     @Override
-    public Set<MemoryModuleType<?>> requires() {
+    public synchronized Set<MemoryModuleType<?>> requires() {
         return Sets.union(super.requires(), Set.of(MemoryModuleType.HAS_HUNTING_COOLDOWN));
     }
 }
