--- a/net/minecraft/world/entity/ai/sensing/NearestBedSensor.java
+++ b/net/minecraft/world/entity/ai/sensing/NearestBedSensor.java
@@ -18,6 +_,7 @@
 import net.minecraft.world.entity.ai.village.poi.PoiType;
 import net.minecraft.world.entity.ai.village.poi.PoiTypes;
 import net.minecraft.world.level.pathfinder.Path;
+import net.neoforged.neoforge.server.region.RegionizedTickContext;
 
 public class NearestBedSensor extends Sensor<Mob> {
     private static final int CACHE_TIMEOUT = 40;
@@ -36,10 +_,10 @@
         return ImmutableSet.of(MemoryModuleType.NEAREST_BED);
     }
 
-    protected void doTick(ServerLevel level, Mob body) {
+    protected synchronized void doTick(ServerLevel level, Mob body) {
         if (body.isBaby()) {
             this.triedCount = 0;
-            this.lastUpdate = level.getGameTime() + level.getRandom().nextInt(20);
+            this.lastUpdate = level.getGameTime() + RegionizedTickContext.currentRandom().orElse(level.getRandom()).nextInt(20);
             PoiManager poiManager = level.getPoiManager();
             Predicate<BlockPos> cacheTest = pos -> {
                 long key = pos.asLong();
