--- a/net/minecraft/world/entity/ai/sensing/PlayerSensor.java
+++ b/net/minecraft/world/entity/ai/sensing/PlayerSensor.java
@@ -12,6 +_,7 @@
 import net.minecraft.world.entity.ai.attributes.Attributes;
 import net.minecraft.world.entity.ai.memory.MemoryModuleType;
 import net.minecraft.world.entity.player.Player;
+import net.neoforged.neoforge.server.region.RegionizedWorldGuard;
 
 public class PlayerSensor extends Sensor<LivingEntity> {
     @Override
@@ -26,18 +_,19 @@
 
     @Override
     protected void doTick(ServerLevel level, LivingEntity body) {
-        List<Player> players = level.players()
+        List<Player> players = List.copyOf(level.players())
             .stream()
             .filter(EntitySelector.NO_SPECTATORS)
+            .filter(player -> RegionizedWorldGuard.restrictEntityToCurrentRegion(level, player) != null)
             .filter(player -> body.closerThan(player, this.getFollowDistance(body)))
             .sorted(Comparator.comparingDouble(body::distanceToSqr))
             .collect(Collectors.toList());
         Brain<?> brain = body.getBrain();
-        brain.setMemory(MemoryModuleType.NEAREST_PLAYERS, players);
+        brain.setMemory(MemoryModuleType.NEAREST_PLAYERS, List.copyOf(players));
         List<Player> visiblePlayers = players.stream().filter(livingEntity -> isEntityTargetable(level, body, livingEntity)).collect(Collectors.toList());
         brain.setMemory(MemoryModuleType.NEAREST_VISIBLE_PLAYER, visiblePlayers.isEmpty() ? null : visiblePlayers.get(0));
         List<Player> visibleAttackablePlayers = visiblePlayers.stream().filter(livingEntity -> isEntityAttackable(level, body, livingEntity)).toList();
-        brain.setMemory(MemoryModuleType.NEAREST_VISIBLE_ATTACKABLE_PLAYERS, visibleAttackablePlayers);
+        brain.setMemory(MemoryModuleType.NEAREST_VISIBLE_ATTACKABLE_PLAYERS, List.copyOf(visibleAttackablePlayers));
         brain.setMemory(MemoryModuleType.NEAREST_VISIBLE_ATTACKABLE_PLAYER, visibleAttackablePlayers.isEmpty() ? null : visibleAttackablePlayers.get(0));
     }
 
