--- a/net/minecraft/world/entity/ai/sensing/Sensor.java
+++ b/net/minecraft/world/entity/ai/sensing/Sensor.java
@@ -9,6 +_,7 @@
 import net.minecraft.world.entity.ai.attributes.Attributes;
 import net.minecraft.world.entity.ai.memory.MemoryModuleType;
 import net.minecraft.world.entity.ai.targeting.TargetingConditions;
+import net.neoforged.neoforge.server.region.RegionizedWorldGuard;
 
 public abstract class Sensor<E extends LivingEntity> {
     private static final int DEFAULT_SCAN_RATE = 20;
@@ -37,15 +_,17 @@
         this(20);
     }
 
-    public void randomlyDelayStart(RandomSource randomSource) {
+    public synchronized void randomlyDelayStart(RandomSource randomSource) {
         this.timeToTick = randomSource.nextInt(this.scanRate);
     }
 
-    public final void tick(ServerLevel level, E body) {
+    public final synchronized void tick(ServerLevel level, E body) {
         if (--this.timeToTick <= 0L) {
             this.timeToTick = this.scanRate;
-            this.updateTargetingConditionRanges(body);
-            this.doTick(level, body);
+            synchronized (Sensor.class) {
+                this.updateTargetingConditionRanges(body);
+                this.doTick(level, body);
+            }
         }
     }
 
@@ -63,13 +_,21 @@
 
     public abstract Set<MemoryModuleType<?>> requires();
 
-    public static boolean isEntityTargetable(ServerLevel level, LivingEntity body, LivingEntity entity) {
+    public static synchronized boolean isEntityTargetable(ServerLevel level, LivingEntity body, LivingEntity entity) {
+        if (RegionizedWorldGuard.restrictEntityToCurrentRegion(level, entity) == null) {
+            return false;
+        }
+
         return body.getBrain().isMemoryValue(MemoryModuleType.ATTACK_TARGET, entity)
             ? TARGET_CONDITIONS_IGNORE_INVISIBILITY_TESTING.test(level, body, entity)
             : TARGET_CONDITIONS.test(level, body, entity);
     }
 
-    public static boolean isEntityAttackable(ServerLevel level, LivingEntity body, LivingEntity target) {
+    public static synchronized boolean isEntityAttackable(ServerLevel level, LivingEntity body, LivingEntity target) {
+        if (RegionizedWorldGuard.restrictEntityToCurrentRegion(level, target) == null) {
+            return false;
+        }
+
         return body.getBrain().isMemoryValue(MemoryModuleType.ATTACK_TARGET, target)
             ? ATTACK_TARGET_CONDITIONS_IGNORE_INVISIBILITY_TESTING.test(level, body, target)
             : ATTACK_TARGET_CONDITIONS.test(level, body, target);
@@ -79,7 +_,11 @@
         return rememberPositives(ticks, (level, target) -> isEntityAttackable(level, body, target));
     }
 
-    public static boolean isEntityAttackableIgnoringLineOfSight(ServerLevel level, LivingEntity body, LivingEntity target) {
+    public static synchronized boolean isEntityAttackableIgnoringLineOfSight(ServerLevel level, LivingEntity body, LivingEntity target) {
+        if (RegionizedWorldGuard.restrictEntityToCurrentRegion(level, target) == null) {
+            return false;
+        }
+
         return body.getBrain().isMemoryValue(MemoryModuleType.ATTACK_TARGET, target)
             ? ATTACK_TARGET_CONDITIONS_IGNORE_INVISIBILITY_AND_LINE_OF_SIGHT.test(level, body, target)
             : ATTACK_TARGET_CONDITIONS_IGNORE_LINE_OF_SIGHT.test(level, body, target);
