--- a/net/minecraft/world/entity/EntityReference.java
+++ b/net/minecraft/world/entity/EntityReference.java
@@ -14,6 +_,7 @@
 import net.minecraft.world.level.entity.UniquelyIdentifyable;
 import net.minecraft.world.level.storage.ValueInput;
 import net.minecraft.world.level.storage.ValueOutput;
+import net.neoforged.neoforge.server.region.RegionizedWorldGuard;
 import org.jetbrains.annotations.Contract;
 import org.jspecify.annotations.Nullable;
 
@@ -76,9 +_,15 @@
     }
 
     public @Nullable StoredEntityType getEntity(Level level, Class<StoredEntityType> clazz) {
-        return Player.class.isAssignableFrom(clazz)
+        StoredEntityType entity = Player.class.isAssignableFrom(clazz)
             ? this.getEntity(level::getPlayerInAnyDimension, clazz)
             : this.getEntity(level::getEntityInAnyDimension, clazz);
+        if (entity instanceof Entity levelEntity) {
+            Entity restricted = RegionizedWorldGuard.restrictEntityToCurrentRegion(level, levelEntity);
+            return restricted != null ? clazz.cast(restricted) : null;
+        }
+
+        return entity;
     }
 
     private @Nullable StoredEntityType resolve(@Nullable UniquelyIdentifyable entity, Class<StoredEntityType> clazz) {
