--- a/net/minecraft/world/inventory/InventoryMenu.java
+++ b/net/minecraft/world/inventory/InventoryMenu.java
@@ -66,13 +_,13 @@
             }
 
             @Override
-            public void setByPlayer(ItemStack itemStack, ItemStack previous) {
+            public synchronized void setByPlayer(ItemStack itemStack, ItemStack previous) {
                 owner.onEquipItem(EquipmentSlot.OFFHAND, previous, itemStack);
                 super.setByPlayer(itemStack, previous);
             }
 
             @Override
-            public Identifier getNoItemIcon() {
+            public synchronized Identifier getNoItemIcon() {
                 return InventoryMenu.EMPTY_ARMOR_SLOT_SHIELD;
             }
         });
@@ -83,14 +_,14 @@
     }
 
     @Override
-    public void slotsChanged(Container container) {
+    public synchronized void slotsChanged(Container container) {
         if (this.owner.level() instanceof ServerLevel level) {
             CraftingMenu.slotChangedCraftingGrid(this, level, this.owner, this.craftSlots, this.resultSlots, null);
         }
     }
 
     @Override
-    public void removed(Player player) {
+    public synchronized void removed(Player player) {
         super.removed(player);
         this.resultSlots.clearContent();
         if (!player.level().isClientSide()) {
@@ -104,7 +_,7 @@
     }
 
     @Override
-    public ItemStack quickMoveStack(Player player, int slotIndex) {
+    public synchronized ItemStack quickMoveStack(Player player, int slotIndex) {
         ItemStack clicked = ItemStack.EMPTY;
         Slot slot = this.slots.get(slotIndex);
         if (slot.hasItem()) {
@@ -166,7 +_,7 @@
     }
 
     @Override
-    public boolean canTakeItemForPickAll(ItemStack carried, Slot target) {
+    public synchronized boolean canTakeItemForPickAll(ItemStack carried, Slot target) {
         return target.container != this.resultSlots && super.canTakeItemForPickAll(carried, target);
     }
 
