--- a/net/minecraft/world/entity/vehicle/minecart/MinecartHopper.java
+++ b/net/minecraft/world/entity/vehicle/minecart/MinecartHopper.java
@@ -19,6 +_,8 @@
 import net.minecraft.world.level.block.state.properties.RailShape;
 import net.minecraft.world.level.storage.ValueInput;
 import net.minecraft.world.level.storage.ValueOutput;
+import net.minecraft.world.phys.AABB;
+import net.neoforged.neoforge.server.region.RegionizedWorldGuard;
 
 public class MinecartHopper extends AbstractMinecartContainer implements Hopper {
     private static final boolean DEFAULT_ENABLED = true;
@@ -105,10 +_,12 @@
         if (HopperBlockEntity.suckInItems(this.level(), this)) {
             return true;
         } else {
-            for (ItemEntity entity : this.level()
-                .getEntitiesOfClass(ItemEntity.class, this.getBoundingBox().inflate(0.25, 0.0, 0.25), EntitySelector.ENTITY_STILL_ALIVE)) {
-                if (HopperBlockEntity.addItem(this, entity)) {
-                    return true;
+            AABB pickupBox = RegionizedWorldGuard.restrictToCurrentRegion(this.level(), this.getBoundingBox().inflate(0.25, 0.0, 0.25));
+            if (pickupBox != null) {
+                for (ItemEntity entity : this.level().getEntitiesOfClass(ItemEntity.class, pickupBox, EntitySelector.ENTITY_STILL_ALIVE)) {
+                    if (HopperBlockEntity.addItem(this, entity)) {
+                        return true;
+                    }
                 }
             }
 
