--- a/net/minecraft/world/entity/animal/frog/ShootTongue.java
+++ b/net/minecraft/world/entity/animal/frog/ShootTongue.java
@@ -51,7 +_,12 @@
     }
 
     protected boolean checkExtraStartConditions(ServerLevel level, Frog body) {
-        LivingEntity target = body.getBrain().getMemory(MemoryModuleType.ATTACK_TARGET).get();
+        LivingEntity target = BehaviorUtils.regionSafeEntity(body, body.getBrain().getMemory(MemoryModuleType.ATTACK_TARGET).get());
+        if (target == null) {
+            body.getBrain().eraseMemory(MemoryModuleType.ATTACK_TARGET);
+            return false;
+        }
+
         boolean canPathfindToTarget = this.canPathfindToTarget(body, target);
         if (!canPathfindToTarget) {
             body.getBrain().eraseMemory(MemoryModuleType.ATTACK_TARGET);
@@ -63,12 +_,19 @@
 
     protected boolean canStillUse(ServerLevel level, Frog body, long timestamp) {
         return body.getBrain().hasMemoryValue(MemoryModuleType.ATTACK_TARGET)
+            && BehaviorUtils.regionSafeEntity(body, body.getBrain().getMemory(MemoryModuleType.ATTACK_TARGET).orElse(null)) != null
             && this.state != ShootTongue.State.DONE
             && !body.getBrain().hasMemoryValue(MemoryModuleType.IS_PANICKING);
     }
 
     protected void start(ServerLevel level, Frog body, long timestamp) {
-        LivingEntity target = body.getBrain().getMemory(MemoryModuleType.ATTACK_TARGET).get();
+        LivingEntity target = BehaviorUtils.regionSafeEntity(body, body.getBrain().getMemory(MemoryModuleType.ATTACK_TARGET).get());
+        if (target == null) {
+            body.getBrain().eraseMemory(MemoryModuleType.ATTACK_TARGET);
+            this.state = ShootTongue.State.DONE;
+            return;
+        }
+
         BehaviorUtils.lookAtEntity(body, target);
         body.setTongueTarget(target);
         body.getBrain().setMemory(MemoryModuleType.WALK_TARGET, new WalkTarget(target.position(), 2.0F, 0));
@@ -97,7 +_,13 @@
     }
 
     protected void tick(ServerLevel level, Frog body, long timestamp) {
-        LivingEntity target = body.getBrain().getMemory(MemoryModuleType.ATTACK_TARGET).get();
+        LivingEntity target = BehaviorUtils.regionSafeEntity(body, body.getBrain().getMemory(MemoryModuleType.ATTACK_TARGET).orElse(null));
+        if (target == null) {
+            body.getBrain().eraseMemory(MemoryModuleType.ATTACK_TARGET);
+            this.state = ShootTongue.State.DONE;
+            return;
+        }
+
         body.setTongueTarget(target);
         switch (this.state) {
             case MOVE_TO_TARGET:
