--- a/net/minecraft/world/level/block/EnderChestBlock.java
+++ b/net/minecraft/world/level/block/EnderChestBlock.java
@@ -7,6 +_,7 @@
 import net.minecraft.network.chat.Component;
 import net.minecraft.server.level.ServerLevel;
 import net.minecraft.stats.Stats;
+import net.neoforged.neoforge.server.region.RegionizedWorldGuard;
 import net.minecraft.util.RandomSource;
 import net.minecraft.world.InteractionResult;
 import net.minecraft.world.SimpleMenuProvider;
@@ -88,7 +_,7 @@
                     player.openMenu(
                         new SimpleMenuProvider((containerId, inventory, p) -> ChestMenu.threeRows(containerId, inventory, container), CONTAINER_TITLE)
                     );
-                    player.awardStat(Stats.OPEN_ENDERCHEST);
+                    RegionizedWorldGuard.awardStat(player, Stats.OPEN_ENDERCHEST);
                     PiglinAi.angerNearbyPiglins(serverLevel, player, true);
                 }
 
