--- a/net/minecraft/world/entity/boss/enderdragon/phases/DragonLandingPhase.java
+++ b/net/minecraft/world/entity/boss/enderdragon/phases/DragonLandingPhase.java
@@ -47,7 +_,7 @@
     }
 
     @Override
-    public void doServerTick(ServerLevel level) {
+    public synchronized void doServerTick(ServerLevel level) {
         if (this.targetLocation == null) {
             this.targetLocation = Vec3.atBottomCenterOf(
                 level.getHeightmapPos(Heightmap.Types.MOTION_BLOCKING_NO_LEAVES, EndPodiumFeature.getLocation(this.dragon.getFightOrigin()))
@@ -73,12 +_,12 @@
     }
 
     @Override
-    public void begin() {
+    public synchronized void begin() {
         this.targetLocation = null;
     }
 
     @Override
-    public @Nullable Vec3 getFlyTargetLocation() {
+    public synchronized @Nullable Vec3 getFlyTargetLocation() {
         return this.targetLocation;
     }
 
