--- a/net/minecraft/world/entity/animal/nautilus/AbstractNautilus.java
+++ b/net/minecraft/world/entity/animal/nautilus/AbstractNautilus.java
@@ -57,6 +_,7 @@
 import net.minecraft.world.level.pathfinder.PathType;
 import net.minecraft.world.phys.Vec2;
 import net.minecraft.world.phys.Vec3;
+import net.neoforged.neoforge.server.region.RegionizedWorldGuard;
 import org.jspecify.annotations.Nullable;
 
 public abstract class AbstractNautilus extends TamableAnimal implements PlayerRideableJumping, HasCustomInventoryScreen {
@@ -136,10 +_,18 @@
     ) {
         int seaLevel = level.getSeaLevel();
         int minSpawnLevel = seaLevel - 25;
+        BlockPos below = pos.below();
+        BlockPos above = pos.above();
         return pos.getY() >= minSpawnLevel
             && pos.getY() <= seaLevel - 5
-            && level.getFluidState(pos.below()).is(FluidTags.WATER)
-            && level.getBlockState(pos.above()).is(Blocks.WATER);
+            && canReadBlock(level, below)
+            && canReadBlock(level, above)
+            && level.getFluidState(below).is(FluidTags.WATER)
+            && level.getBlockState(above).is(Blocks.WATER);
+    }
+
+    private static boolean canReadBlock(LevelAccessor level, BlockPos pos) {
+        return !(level instanceof Level levelInstance) || RegionizedWorldGuard.shouldLoadChunk(levelInstance, pos.getX() >> 4, pos.getZ() >> 4);
     }
 
     @Override
@@ -445,8 +_,9 @@
     @Override
     public boolean hurtServer(ServerLevel level, DamageSource source, float damage) {
         boolean wasHurt = super.hurtServer(level, source, damage);
-        if (wasHurt && source.getEntity() instanceof LivingEntity sourceEntity) {
-            NautilusAi.setAngerTarget(level, this, sourceEntity);
+        Entity sourceEntity = RegionizedWorldGuard.restrictEntityToCurrentRegion(this.level(), source.getEntity());
+        if (wasHurt && sourceEntity instanceof LivingEntity livingSource) {
+            NautilusAi.setAngerTarget(level, this, livingSource);
         }
 
         return wasHurt;
@@ -461,7 +_,7 @@
     public SpawnGroupData finalizeSpawn(
         ServerLevelAccessor level, DifficultyInstance difficulty, EntitySpawnReason spawnReason, @Nullable SpawnGroupData groupData
     ) {
-        RandomSource random = level.getRandom();
+        RandomSource random = this.getRandom();
         NautilusAi.initMemories(this, random);
         return super.finalizeSpawn(level, difficulty, spawnReason, groupData);
     }
