--- a/net/minecraft/world/item/ArmorStandItem.java
+++ b/net/minecraft/world/item/ArmorStandItem.java
@@ -17,6 +_,7 @@
 import net.minecraft.world.level.gameevent.GameEvent;
 import net.minecraft.world.phys.AABB;
 import net.minecraft.world.phys.Vec3;
+import net.neoforged.neoforge.server.region.RegionizedWorldGuard;
 
 public class ArmorStandItem extends Item {
     public ArmorStandItem(Item.Properties properties) {
@@ -35,7 +_,8 @@
             ItemStack itemStack = context.getItemInHand();
             Vec3 pos = Vec3.atBottomCenterOf(blockPos);
             AABB box = EntityType.ARMOR_STAND.getDimensions().makeBoundingBox(pos.x(), pos.y(), pos.z());
-            if (level.noCollision(null, box) && level.getEntities(null, box).isEmpty()) {
+            AABB regionBox = RegionizedWorldGuard.restrictToCurrentRegion(level, box);
+            if (regionBox != null && regionBox.equals(box) && level.noCollision(null, box) && level.getEntities(null, box).isEmpty()) {
                 if (level instanceof ServerLevel serverLevel) {
                     Consumer<ArmorStand> entityConfig = EntityType.createDefaultStackConfig(serverLevel, itemStack, context.getPlayer());
                     ArmorStand entity = EntityType.ARMOR_STAND.create(serverLevel, entityConfig, blockPos, EntitySpawnReason.SPAWN_ITEM_USE, true, true);
