--- a/net/minecraft/world/item/EmptyMapItem.java
+++ b/net/minecraft/world/item/EmptyMapItem.java
@@ -7,6 +_,7 @@
 import net.minecraft.world.InteractionResult;
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.level.Level;
+import net.neoforged.neoforge.server.region.RegionizedWorldGuard;
 
 public class EmptyMapItem extends Item {
     public EmptyMapItem(Item.Properties properties) {
@@ -18,7 +_,7 @@
         ItemStack itemStack = player.getItemInHand(hand);
         if (level instanceof ServerLevel serverLevel) {
             itemStack.consume(1, player);
-            player.awardStat(Stats.ITEM_USED.get(this));
+            RegionizedWorldGuard.awardStat(player, Stats.ITEM_USED.get(this));
             serverLevel.playSound(null, player, SoundEvents.UI_CARTOGRAPHY_TABLE_TAKE_RESULT, player.getSoundSource(), 1.0F, 1.0F);
             ItemStack map = MapItem.create(serverLevel, player.getBlockX(), player.getBlockZ(), (byte)0, true, false);
             if (itemStack.isEmpty()) {
