--- a/net/minecraft/world/item/JukeboxPlayable.java
+++ b/net/minecraft/world/item/JukeboxPlayable.java
@@ -13,6 +_,7 @@
 import net.minecraft.network.chat.Style;
 import net.minecraft.network.codec.StreamCodec;
 import net.minecraft.stats.Stats;
+import net.neoforged.neoforge.server.region.RegionizedWorldGuard;
 import net.minecraft.world.InteractionResult;
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.item.component.TooltipProvider;
@@ -48,7 +_,7 @@
                         level.gameEvent(GameEvent.BLOCK_CHANGE, pos, GameEvent.Context.of(player, state));
                     }
 
-                    player.awardStat(Stats.PLAY_RECORD);
+                    RegionizedWorldGuard.awardStat(player, Stats.PLAY_RECORD);
                 }
 
                 return InteractionResult.SUCCESS;
