--- a/net/minecraft/world/item/LingeringPotionItem.java
+++ b/net/minecraft/world/item/LingeringPotionItem.java
@@ -27,7 +_,7 @@
             SoundEvents.LINGERING_POTION_THROW,
             SoundSource.NEUTRAL,
             0.5F,
-            0.4F / (level.getRandom().nextFloat() * 0.4F + 0.8F)
+            0.4F / (player.getRandom().nextFloat() * 0.4F + 0.8F)
         );
         return super.use(level, player, hand);
     }
