--- a/net/minecraft/world/item/MinecartItem.java
+++ b/net/minecraft/world/item/MinecartItem.java
@@ -14,7 +_,9 @@
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.block.state.properties.RailShape;
 import net.minecraft.world.level.gameevent.GameEvent;
+import net.minecraft.world.phys.AABB;
 import net.minecraft.world.phys.Vec3;
+import net.neoforged.neoforge.server.region.RegionizedWorldGuard;
 
 public class MinecartItem extends Item {
     private final EntityType<? extends AbstractMinecart> type;
@@ -34,7 +_,7 @@
         } else {
             ItemStack itemStack = context.getItemInHand();
             RailShape shape = blockState.getBlock() instanceof BaseRailBlock
-                ? blockState.getValue(((BaseRailBlock)blockState.getBlock()).getShapeProperty())
+                ? ((BaseRailBlock)blockState.getBlock()).getRailDirection(blockState, level, pos, null)
                 : RailShape.NORTH_SOUTH;
             double offset = 0.0;
             if (shape.isSlope()) {
@@ -48,8 +_,14 @@
             if (cart == null) {
                 return InteractionResult.FAIL;
             } else {
+                AABB cartBox = cart.getBoundingBox();
+                AABB regionCartBox = RegionizedWorldGuard.restrictToCurrentRegion(level, cartBox);
+                if (regionCartBox == null || !regionCartBox.equals(cartBox)) {
+                    return InteractionResult.FAIL;
+                }
+
                 if (AbstractMinecart.useExperimentalMovement(level)) {
-                    for (Entity entity : level.getEntities(null, cart.getBoundingBox())) {
+                    for (Entity entity : level.getEntities(null, cartBox)) {
                         if (entity instanceof AbstractMinecart) {
                             return InteractionResult.FAIL;
                         }
