--- a/net/minecraft/world/item/SnowballItem.java
+++ b/net/minecraft/world/item/SnowballItem.java
@@ -12,6 +_,7 @@
 import net.minecraft.world.entity.projectile.Projectile;
 import net.minecraft.world.entity.projectile.throwableitemprojectile.Snowball;
 import net.minecraft.world.level.Level;
+import net.neoforged.neoforge.server.region.RegionizedWorldGuard;
 
 public class SnowballItem extends Item implements ProjectileItem {
     public static final float PROJECTILE_SHOOT_POWER = 1.5F;
@@ -31,13 +_,13 @@
             SoundEvents.SNOWBALL_THROW,
             SoundSource.NEUTRAL,
             0.5F,
-            0.4F / (level.getRandom().nextFloat() * 0.4F + 0.8F)
+            0.4F / (player.getRandom().nextFloat() * 0.4F + 0.8F)
         );
         if (level instanceof ServerLevel serverLevel) {
             Projectile.spawnProjectileFromRotation(Snowball::new, serverLevel, itemStack, player, 0.0F, 1.5F, 1.0F);
         }
 
-        player.awardStat(Stats.ITEM_USED.get(this));
+        RegionizedWorldGuard.awardStat(player, Stats.ITEM_USED.get(this));
         itemStack.consume(1, player);
         return InteractionResult.SUCCESS;
     }
