--- a/net/minecraft/world/effect/MobEffectUtil.java
+++ b/net/minecraft/world/effect/MobEffectUtil.java
@@ -10,6 +_,7 @@
 import net.minecraft.world.entity.Entity;
 import net.minecraft.world.entity.LivingEntity;
 import net.minecraft.world.phys.Vec3;
+import net.neoforged.neoforge.server.region.RegionizedWorldGuard;
 import org.jspecify.annotations.Nullable;
 
 public final class MobEffectUtil {
@@ -53,7 +_,8 @@
     ) {
         Holder<MobEffect> effect = effectInstance.getEffect();
         List<ServerPlayer> players = level.getPlayers(
-            input -> input.gameMode.isSurvival()
+            input -> RegionizedWorldGuard.restrictEntityToCurrentRegion(level, input) != null
+                && input.gameMode.isSurvival()
                 && (source == null || !source.isAlliedTo(input))
                 && position.closerThan(input.position(), radius)
                 && (
