--- a/net/minecraft/world/level/block/entity/TrialSpawnerBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/TrialSpawnerBlockEntity.java
@@ -35,7 +_,7 @@
     }
 
     @Override
-    protected void loadAdditional(ValueInput input) {
+    protected synchronized void loadAdditional(ValueInput input) {
         super.loadAdditional(input);
         this.trialSpawner.load(input);
         if (this.level != null) {
@@ -44,7 +_,7 @@
     }
 
     @Override
-    protected void saveAdditional(ValueOutput output) {
+    protected synchronized void saveAdditional(ValueOutput output) {
         super.saveAdditional(output);
         this.trialSpawner.store(output);
     }
@@ -55,11 +_,13 @@
 
     @Override
     public CompoundTag getUpdateTag(HolderLookup.Provider registries) {
-        return this.trialSpawner.getStateData().getUpdateTag(this.getBlockState().getValue(TrialSpawnerBlock.STATE));
+        synchronized (this.trialSpawner) {
+            return this.trialSpawner.getStateData().getUpdateTag(this.getBlockState().getValue(TrialSpawnerBlock.STATE));
+        }
     }
 
     @Override
-    public void setEntityId(EntityType<?> type, RandomSource random) {
+    public synchronized void setEntityId(EntityType<?> type, RandomSource random) {
         if (this.level == null) {
             Util.logAndPauseIfInIde("Expected non-null level");
         } else {
@@ -68,25 +_,25 @@
         }
     }
 
-    public TrialSpawner getTrialSpawner() {
+    public synchronized TrialSpawner getTrialSpawner() {
         return this.trialSpawner;
     }
 
     @Override
-    public TrialSpawnerState getState() {
+    public synchronized TrialSpawnerState getState() {
         return !this.getBlockState().hasProperty(BlockStateProperties.TRIAL_SPAWNER_STATE)
             ? TrialSpawnerState.INACTIVE
             : this.getBlockState().getValue(BlockStateProperties.TRIAL_SPAWNER_STATE);
     }
 
     @Override
-    public void setState(Level level, TrialSpawnerState state) {
+    public synchronized void setState(Level level, TrialSpawnerState state) {
         this.setChanged();
         level.setBlockAndUpdate(this.worldPosition, this.getBlockState().setValue(BlockStateProperties.TRIAL_SPAWNER_STATE, state));
     }
 
     @Override
-    public void markUpdated() {
+    public synchronized void markUpdated() {
         this.setChanged();
         if (this.level != null) {
             this.level.sendBlockUpdated(this.worldPosition, this.getBlockState(), this.getBlockState(), 3);
