--- a/net/minecraft/server/commands/PardonIpCommand.java
+++ b/net/minecraft/server/commands/PardonIpCommand.java
@@ -28,6 +_,10 @@
     }
 
     private static int unban(CommandSourceStack source, String ip) throws CommandSyntaxException {
+        if (RegionizedCommandHelper.scheduleGlobalIfRegionized(source, () -> unban(source, ip))) {
+            return 1;
+        }
+
         if (!InetAddresses.isInetAddress(ip)) {
             throw ERROR_INVALID.create();
         } else {
